ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,18 +702,89 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.singleton(newId), expectedEndToken);
+
+    // same test as above, but with old incarnationId
+    bytesRead = log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on startup after recovery with old incarnationId
+    startToken = new StoreFindToken(secondRecordFileSpan.getEndOffset(), oldSessionId, oldIncarnationId, false);
+    // token should get reset internally, returning all the keys and the returned token should be correct (offset in
+    // it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // create a token that is not past the index end offset on startup after recovery, but with old incarnationId.
+    // again the token should be reset internally, all keys should be returned and the returned token should be correct
+    // (offset in it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, oldIncarnationId, false);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // test cases where no new puts/deletes have gone in after restart.
+    // after restart, replica 1(assuming no new puts have gone in) will return a journal based token referring to
+    // logEndOffsetOnStartUp to replica2.
+    // On the subsequent findEntriesSince() call, the index should start returning entries starting from that offset
+    // and should not consider it as non-inclusive offset
+    recovery = new MessageStoreRecovery() {
+      @Override
+      public List<MessageInfo> recover(Read read, long startOffset, long endOffset, StoreKeyFactory factory)
+          throws IOException {
+        return Collections.emptyList();
+      }
+    };
+    reloadIndex(true);
+    lastRecordOffset = index.journal.getLastOffset();
+    bytesRead = log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on start up after restart.
+    startToken = new StoreFindToken(
+        new Offset(log.getEndOffset().getName(), (lastRecordOffset.getOffset() + 2 * PUT_RECORD_SIZE)), oldSessionId,
```

Comment:
I think you should just write this as a new test. This test is getting a little too long and has too much state.